src/utils/database.ts

```diff
@@ -5434,6 +5591,262 @@ const recordOffRampTransaction = async (event: OnrampMoneyWebhook) => {
+      .select(
+        'id, created_at, owner_account_address, default_chain_id, notification, pin_hash'
+      )
+      .single()
+
+    if (error) {
+      console.error('Database error in updatePaymentPreferences:', error)
+      Sentry.captureException(error)
+      throw new Error('Could not update payment preferences')
+    }
+
+    if (!result) {
+      throw new Error('No data returned from database operation')
+    }
+
+    const { pin_hash, ...rest } = result
+    return {
+      ...rest,
+      hasPin: !!pin_hash,
+    }
+  } catch (error) {
+    console.error('Unexpected error in updatePaymentPreferences:', error)
+    Sentry.captureException(error)
+    throw error instanceof Error
+      ? error
+      : new Error('Could not update payment preferences')
+  }
+}
+
+const createPinHash = async (pin: string): Promise<string> => {
+  const pinWithSalt = `${pin}${PIN_SALT}`
+  return await argon2.hash(pinWithSalt, {
+    type: argon2.argon2id,
+    memoryCost: 2 ** 16,
+    timeCost: 3,
+    parallelism: 1,
+  })
+}
+
+const verifyUserPin = async (
+  owner_account_address: string,
+  pin: string
+): Promise<boolean> => {
+  try {
+    // Get raw data including pin_hash for verification
+    const { data, error } = await db.supabase
+      .from('payment_preferences')
+      .select('pin_hash')
+      .eq('owner_account_address', owner_account_address.toLowerCase())
+      .single()
+
+    if (error) {
+      console.error('Error fetching PIN for verification:', error)
+      return false
+    }
+
+    if (!data?.pin_hash) {
+      return false
+    }
+
+    const pinWithSalt = `${pin}${PIN_SALT}`
+    const isValid = await argon2.verify(data.pin_hash, pinWithSalt)
+    return isValid
+  } catch (error) {
+    console.error('Unexpected error in verifyUserPin:', error)
+    return false
+  }
+}
+
+const createVerification = async (
+  owner_account_address: string,
+  code: string,
+  channel: VerificationChannel,
+  expiresAt: Date
+): Promise<void> => {
+  try {
+    const { error } = await db.supabase.from('verifications').insert({
+      owner_account_address: owner_account_address,
+      code_hash: await createPinHash(code),
+      channel: channel,
+      expires_at: expiresAt.toISOString(),
+    })
+
+    if (error) {
+      console.error('Error creating verification:', error)
+      throw new Error('Failed to create verification')
+    }
+  } catch (error) {
+    console.error('Error in createVerification:', error)
+    throw error
+  }
+}
+
+const verifyVerificationCode = async (
+  owner_account_address: string,
+  code: string,
+  channel: VerificationChannel
+): Promise<boolean> => {
+  try {
+    const { data, error } = await db.supabase
+      .from('verifications')
+      .select('*')
+      .eq('owner_account_address', owner_account_address)
+      .eq('channel', channel)
+      .gte('expires_at', new Date().toISOString())
+      .order('created_at', { ascending: false })
+      .limit(1)
```

Comment:
what if multiple exists ? say the user generated 2 and used the first one he/she generated?